src/pip/_internal/index.py

```diff
@@ -219,13 +232,37 @@ def _get_html_page(link, session=None):
         reason += str(exc)
         _handle_get_page_fail(link, reason, url, meth=logger.info)
     except requests.ConnectionError as exc:
-        _handle_get_page_fail(link, "connection error: %s" % exc, url)
+        # connection failures are not retried
+        _handle_get_page_fail(
+            link,
+            "connection error: %s" % exc,
+            url,
+            logger.warning,
+        )
+        unresponsive_hosts and unresponsive_hosts.track(netloc)
     except requests.Timeout:
         _handle_get_page_fail(link, "timed out", url)
     else:
         return HTMLPage(resp.content, resp.url, resp.headers)
 
 
+class HostTracker(object):
```

Comment:
This class can be just be replaced with a set instead, since it isn't doing anything except for a single action on the set in each of it's methods.

Reply:
Hi @pradyunsg , I'd used this style (against my instinct) based on the previous pr comments "a new class can be used to handle this" and "session.is_host_blacklisted(netloc)"/"session.add_host_to_blacklist(netloc)" which, I think, were based on the idea of abstracting the interface from the data storage? If you're sure it would be ok to remove the `HostTracker` and just have a set in the Finder it would indeed be exactly the same number of lines in the finder class - shall I change it then?

Reply:
That particular comment was meant to be an alternate proposal to consider. I would be fine if this is simply a set.

Reply:
I do prefer just a plain set.

While abstraction is a good idea, I think the fact that each method of the class is a wrapper of a single set operation, suggests that dropping it would reduce complexity (since the variant without abstraction would still be similar in terms of readability).

PS - I've not looked at the PR due to time constraints, thanks @uranusjr for taking the time to review this and providing valuable feedback.